### PR TITLE
[FlexCounter] Refactor FlexCounter class 

### PR DESCRIPTION
**Pull request by Junchao-Mellanox**
HLD: https://github.com/sonic-net/SONiC/pull/1020

**Why I did it?**
FlexCounter class is a representation of flex counter group which supports querying multiple types of statistic/attributes. It supports multiple statistic/attribute types such as port counter, port debug counter, queue counter, queue attribute and so on. For each statistic/attribute type, it defines several member functions:

- setXXXCounterList: e.g. setPortCounterList, setPortDebugCounterList
- removeXXX: e.g. removePort, removeQueue
- collectXXXCounters: e.g. collectPortCounters, collectQueueCounters
- collectXXXAttr: e.g. collectQueueAttrs, collectPriorityGroupAttrs
- so on
- 
For different statistic/attribute types, these functions have very similar logic. This PR moved similar logic to single place to avoid redundant code.

**How I test it?**

Almost full unit test coverage of newly added code.

### CODE REVIEW

**Comment by Junchao-Mellanox** [comment]
hi @kcudnik, could you please help review? The build failure is due to that Azure pipeline is using buster and GCC 8 to build sonic-sairedis, I have a PR to address this: https://github.com/Azure/sonic-sairedis/pull/1069

**Comment by kcudnik** [comment]
i added comment on other PR

**Comment by Junchao-Mellanox** [comment]
Hi @kcudnik , could you please review this? All checkers have passed.

**Comment by Junchao-Mellanox** [comment]
Hi @keboliu and @stephenxs , could you please review and sign-off?

**Review by keboliu — DISMISSED** [review]

**Comment by Junchao-Mellanox** [comment]
Hi @kcudnik , kindly reminder...

**Review by stephenxs — DISMISSED** [review]

**Review by kcudnik — DISMISSED** [review]
i'm approving this since this PR is too huge to understand what is chaning here, im assuming that this could cause some issues and maybe introduce some new bugs, since changing so much code is unsafe :/

File: syncd/FlexCounter.h [hunk 1]
```diff
@@ -13,17 +13,55 @@ extern "C" {
...
+    class BaseCounterContext
+    {
+    public:
+        BaseCounterContext(const std::string &name);
+        void addPlugins(
+            _In_ const std::vector<std::string>& shaStrings);
+
+        bool hasPlugin() const {return !m_plugins.empty();}
+
+        void removePlugins() {m_plugins.clear();}
```
[the remark this answers is not in the document]
Comment (Junchao-Mellanox) [reply syncd/FlexCounter.h:29]: `removePlugins` is called inside `removeCounterPlugins` which already under mutex.

**Review by keboliu — APPROVED** [review]

**Review by kcudnik — APPROVED** [review]

**Review by stephenxs — APPROVED** [review]

**Comment by Junchao-Mellanox** [comment]
Hi @kcudnik , all reviewers have approved this PR, could you please help merge it?

**Pull request merged**